sys/net/gnrc/link_layer/csma_sender/gnrc_csma_sender.c

```diff
@@ -0,0 +1,245 @@
+{
+    if (be < macMinBE) {
+        be = macMinBE;
+    }
+    if (be > macMaxBE) {
+        be = macMaxBE;
+    }
+    uint32_t max_backoff = ((1 << be) - 1) * A_UNIT_BACKOFF_PERIOD_uS;
+
+    uint32_t period = genrand_uint32() % max_backoff;
+    if (period < A_UNIT_BACKOFF_PERIOD_uS) {
+        period = A_UNIT_BACKOFF_PERIOD_uS;
+    }
+
+    return period;
+}
+
+
+/**
+ * @brief Perform a CCA and send the given packet if medium is available
+ *
+ * @param[in] device    netdev device, needs to be already initialized
+ * @param[in] data      pointer to the data in the packet buffer
+ *
+ * @return              the return value of device driver's @c send_data()
+ *                      function if medium was avilable
+ * @return              -EBUSY if radio medium was not available
+ *                      to send the given data
+ */
+static int send_if_cca(gnrc_netdev_t *device, gnrc_pktsnip_t *data)
+{
+    netopt_enable_t hwfeat;
+
+    /* perform a CCA */
+    DEBUG("Checking radio medium availability...\n");
+    int res = device->driver->get(dev,
+                                  NETOPT_IS_CHANNEL_CLR,
+                                  (void *) &hwfeat,
+                                  sizeof(netopt_enable_t));
+    if (res < 0) {
+        /* normally impossible: we got a big internal problem! */
+        DEBUG("!!! DEVICE DRIVER FAILURE! TRANSMISSION ABORTED!\n");
+        return -ECANCELED;
+    }
+
+    /* if medium is clear, send the packet and return */
+    if (hwfeat == NETOPT_ENABLE) {
+        DEBUG("Radio medium available: sending packet.\n");
+        return dev->driver->send_data(device, data);
+    }
+
+    /* if we arrive here, medium was not available for transmission */
+    DEBUG("Radio medium busy.\n");
+    return -EBUSY;
+}
+
+/*------------------------- "EXPORTED" FUNCTIONS -------------------------*/
+
+void set_csma_mac_min_be(uint8_t val)
+{
+    macMinBE = val;
+}
+
+void set_csma_mac_max_be(uint8_t val)
+{
+    macMaxBE = val;
+}
+
+void set_csma_mac_max_csma_backoffs(uint8_t val);
+{
+    macMaxCSMABackoffs = val;
+}
+
+
+int csma_ca_send(gnrc_netdev_t *dev, gnrc_pktsnip_t *pkt);
+{
+    netopt_enable_t hwfeat;
+
+    /* Does the transceiver do automatic CSMA/CA when sending? */
+    int res = dev->driver->get(dev,
+                               NETOPT_CSMA,
+                               (void *) &hwfeat,
+                               sizeof(netopt_enable_t));
+    bool ok = false;
+    switch (res) {
+    case -ENODEV:
+        /* invalid device pointer given */
+        return -ENODEV;
+    case -ENOTSUP:
+        /* device doesn't make auto-CSMA/CA */
+        ok = false;
+        break;
+    case -EOVERFLOW:  /* (normally impossible...*/
+    case -ECANCELED:
+        DEBUG("!!! DEVICE DRIVER FAILURE! TRANSMISSION ABORTED!\n");
+        /* internal driver error! */
+        return -ECANCELED;
+    default:
+        ok = (hwfeat == NETOPT_ENABLE);
+    }
```

Comment:
Well:
* `ENOTSUP` is not an error (it means we must do the job manually); with your new code, it would be returned as such.
* for `EOVERFLOW` (which should NEVER happen here) and `ECANCELED`, which are internal (severe errors) a debug message is especially printed before returning `ECANCELED` for both, meaning that our code is somehow flawed; I think it deserves a special attention.